akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -277,7 +277,10 @@ private[akka] class QueueSink[T]() extends GraphStageWithMaterializedValue[SinkS
             promise.failure(new IllegalStateException("You have to wait for previous future to be resolved to send another request"))
           case None ⇒
             if (buffer.isEmpty) currentRequest = Some(promise)
-            else sendDownstream(promise)
+            else {
+              if (buffer.used == maxBuffer - 1) tryPull(in)
```

Comment:
Indeed, the allocated buffer is one too big, wasting one element but also using modulo instead of masking for the internal management (due to not being a power of two in size).